src/data/error_model.rs

```diff
@@ -55,81 +57,141 @@ impl ErrorPoly {
     }
 }
 
-impl From<Vec<ErrorModel>> for ErrorModels {
-    fn from(models: Vec<ErrorModel>) -> Self {
-        Self(models)
-    }
-}
-
 /// Collection of error models for all possible outputs in the model/dataset
 /// This struct holds a vector of error models, each corresponding to a specific output
 /// in the pharmacometric analysis.
 ///
 /// This is a wrapper around a vector of [ErrorModel]s, its size is determined by the number of outputs in the model/dataset.
 
-pub struct ErrorModels(Vec<ErrorModel>);
+pub struct ErrorModels {
+    models: Vec<(usize, ErrorModel)>,
+}
+
+impl Default for ErrorModels {
+    fn default() -> Self {
+        Self { models: Vec::new() }
+    }
+}
 
 impl ErrorModels {
-    /// Returns the number of error models in the collection.
-    pub fn len(&self) -> usize {
-        self.0.len()
+    /// Create a new collection of [ErrorModels]
+    pub fn new() -> Self {
+        Self::default()
     }
 
-    /// Returns a vector with the lambda or gamma parameters for each error model.
-    pub fn get_gamlams(&self) -> Vec<f64> {
-        self.0
-            .iter()
-            .map(|model| match model {
-                ErrorModel::Additive { lambda, .. } => *lambda,
-                ErrorModel::Proportional { gamma, .. } => *gamma,
-            })
-            .collect()
+    /// Add an error model for a specific output equation.
+    pub fn add(mut self, outeq: usize, model: ErrorModel) -> Result<Self, ErrorModelError> {
+        if self.models.iter().any(|(id, _)| *id == outeq) {
+            return Err(ErrorModelError::ExistingOutputEquation(outeq));
+        }
+        self.models.push((outeq, model));
+        Ok(self)
     }
 
-    /// Returns the error polynomial associated with the specified output equation.
+    /// Returns the number of error models in the collection.
+    pub fn len(&self) -> usize {
+        self.models.len()
+    }
+    /// Calculates a hash of the ErrorModels collection
+    pub fn hash(&self) -> u64 {
+        let mut hasher = DefaultHasher::new();
+
+        // Sort by output equation index to ensure consistent hashing
+        let mut sorted_models = self.models.clone();
+        sorted_models.sort_by_key(|(outeq, _)| *outeq);
+
+        for (outeq, model) in sorted_models {
+            outeq.hash(&mut hasher);
+            // Hash the discriminant and fields of the ErrorModel enum
+            match model {
+                ErrorModel::Additive { lambda, poly } => {
+                    0u8.hash(&mut hasher); // discriminant for Additive
+                    lambda.to_bits().hash(&mut hasher);
+                    poly.coefficients().0.to_bits().hash(&mut hasher);
```

Comment:
I don't think the coefficients have to be part of the hash calculation as they do not change per user/outeq